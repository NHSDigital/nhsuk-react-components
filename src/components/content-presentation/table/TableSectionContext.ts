import { createContext } from 'react';

export enum TableSection {
  NONE,
  HEAD,
  BODY,
}

const TableSectionContext = createContext<TableSection>(TableSection.NONE);

export default TableSectionContext;
