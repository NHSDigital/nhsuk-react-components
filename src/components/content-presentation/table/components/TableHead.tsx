import React, { FC, HTMLProps } from 'react';
import classNames from 'classnames';
import TableSectionContext, { TableSection } from '../TableSectionContext';

const TableHead: FC<HTMLProps<HTMLTableSectionElement>> = ({ className, children, ...rest }) => (
  <thead className={classNames('nhsuk-table__head', className)} {...rest}>
    <TableSectionContext.Provider value={TableSection.HEAD}>
      {children}
    </TableSectionContext.Provider>
  </thead>
);

TableHead.displayName = 'Table.Head';

export default TableHead;
