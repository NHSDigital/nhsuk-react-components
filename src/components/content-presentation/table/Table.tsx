import React, { ComponentProps, HTMLProps, ReactNode } from 'react';
import classNames from 'classnames';
import TableBody from './components/TableBody';
import TableCaption from './components/TableCaption';
import TableCell from './components/TableCell';
import TableContainer from './components/TableContainer';
import TableHead from './components/TableHead';
import TableRow from './components/TableRow';
import TablePanel from './components/TablePanel';
import TableContext, { ITableContext } from './TableContext';

interface TableProps extends HTMLProps<HTMLTableElement> {
  responsive?: boolean;
  caption?: ReactNode;
  captionProps?: ComponentProps<typeof TableCaption>;
}

interface TableState {
  headings: string[];
}

class Table extends React.PureComponent<TableProps, TableState> {
  static defaultProps = {
    responsive: false,
  };

  static Container = TableContainer;

  static Head = TableHead;

  static Row = TableRow;

  static Cell = TableCell;

  static Body = TableBody;

  static Panel = TablePanel;

  constructor(props: TableProps) {
    super(props);
    this.state = {
      headings: [],
    };
  }

  setHeadings = (headings: string[]): void => {
    const isEqual = headings.reduce(
      (prevValue, heading, index) => prevValue && heading === this.state.headings[index],
      true,
    );

    if (!isEqual) this.setState({ headings });
  };

  render(): JSX.Element {
    const { className, responsive, children, caption, captionProps, ...rest } = this.props;

    const contextValue: ITableContext = {
      isResponsive: Boolean(responsive),
      headings: this.state.headings,
      setHeadings: this.setHeadings,
    };

    return (
      <TableContext.Provider value={contextValue}>
        <table
          className={classNames(
            { 'nhsuk-table': !responsive },
            { 'nhsuk-table-responsive': responsive },
            className,
          )}
          {...rest}
        >
          {caption && <TableCaption {...captionProps}>{caption}</TableCaption>}
          {children}
        </table>
      </TableContext.Provider>
    );
  }
}

export default Table;
