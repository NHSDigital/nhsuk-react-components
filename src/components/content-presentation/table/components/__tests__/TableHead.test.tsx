import { mount, shallow } from 'enzyme';
import React, { useContext } from 'react';
import Table from '../../Table';
import TableSectionContext, { TableSection } from '../../TableSectionContext';
import TableHead from '../TableHead';

describe('Table.Head', () => {
  it('matches snapshot', () => {
    const wrapper = shallow(<TableHead />);
    expect(wrapper).toMatchSnapshot();
    wrapper.unmount();
  });

  it('exposes TableSectionContext', () => {
    let tableSection: TableSection = TableSection.NONE;

    const TestComponent = () => {
      const tableContext = useContext(TableSectionContext);

      if (tableSection !== tableContext) {
        tableSection = tableContext;
      }

      return null;
    };

    const wrapper = mount(
      <Table>
        <TableHead>
          <TestComponent />
        </TableHead>
      </Table>,
    );

    expect(tableSection).toBe(TableSection.HEAD);
    wrapper.unmount();
  });
});
