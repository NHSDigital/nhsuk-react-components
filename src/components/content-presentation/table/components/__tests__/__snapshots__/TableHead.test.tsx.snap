// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table.Head matches snapshot 1`] = `
<thead
  className="nhsuk-table__head"
>
  <ContextProvider
    value={1}
  />
</thead>
`;
